rust/arrow/src/datatypes.rs

```diff
@@ -1500,6 +1590,7 @@ impl Field {
     }
 
     /// Merge field into self if it is compatible. Struct will be merged recursively.
+    /// NOTE: `self` may be updated to unexpected state in case of merge failure.
```

Comment:
I think this is fine, I'm happy with someone who encounters an issue here in future, providing an alternative implementation that doesn't partially mutate `&mut self` if there's a failure.

@houqp any opinion here, as you contributed this function?

Reply:
I think this "good now, better later" strategy is a good one. I agree with @nevi-me 's suggestion

Reply:
yeah, i agree as well. would be better to have an atomic merge implementation done as a separate PR in the future.